src/components/shared/chart/Viewport.js

```diff
@@ -297,21 +292,23 @@ export const withChartViewport: WithChartViewport<*, *> =
         ) {
           this.setState(this.getDefaultState(newProps));
           this._setSizeNextFrame();
-        } else if (
-          this.props.viewportProps.previewSelection !==
-            newProps.viewportProps.previewSelection ||
-          this.props.viewportProps.timeRange !==
-            newProps.viewportProps.timeRange
-        ) {
-          const { previewSelection, timeRange } = newProps.viewportProps;
+        }
+      }
+
+      horizontalViewportMemoized = memoize(
+        (props: ViewportProps) => {
+          const { previewSelection, timeRange } = props.viewportProps;
           const horizontalViewport = this.getHorizontalViewport(
             previewSelection,
             timeRange
           );
-          this.setState({
-            horizontalViewport,
-          });
-        } else if (
+          return horizontalViewport;
+        },
+        { cache: new NamedTupleMap() }
+      );
```

Comment:
I move the piece of code where it calculates `horizontalViewport`, except the `setState()`. 
`case` is set as `NamedTupleMap`, since [the documentation](https://github.com/memoize-immutable/memoize-immutable) of memoize-immutable says: 
> [if] The function accepts a single object of named argument -> use the `NamedTupleMap`.